metals/src/main/scala/scala/meta/internal/worksheets/WorkspaceEditWorksheetPublisher.scala

```diff
@@ -0,0 +1,168 @@
+      editsWithDetails.map(ed => new TextEdit(ed.range, ed.text)).toList
+    val hovers = editsWithDetails.map(
+      ed =>
+        HoverMessage(
+          ed.range
+            .copy(endCharacter = ed.range.getEnd.getCharacter + ed.text.length),
+          ed.details
+        )
+    )
+    val hoverMap = HoverMap(updateWithEdits(source.text, edits), hovers)
+
+    RenderResult(edits, hoverMap)
+  }
+
+  private def publish(
+      path: AbsolutePath,
+      languageClient: MetalsLanguageClient
+  )(rendered: RenderResult): Unit = {
+    hoverMessages = hoverMessages.updated(path, rendered.hovers)
+
+    val params = new ApplyWorkspaceEditParams(
+      new WorkspaceEdit(
+        Map(path.toURI.toString -> (rendered.edits.asJava)).asJava
+      )
+    )
+    languageClient.applyEdit(params)
+  }
+
+  private def renderEdit(
+      statement: EvaluatedWorksheetStatement,
+      source: Input
+  ): RenderEditResult = {
+    val startPosition =
+      new Position(statement.position.endLine, statement.position.endColumn)
+    val endPosition =
+      locatePreviousEdit(statement, source).getOrElse(startPosition)
+
+    RenderEditResult(
+      new Range(startPosition, endPosition),
+      renderMessage(statement),
+      statement.details()
+    )
+  }
+
+  private def renderMessage(statement: EvaluatedWorksheetStatement): String = {
+    import statement._
+    val out = new StringBuilder()
+    out.append("  /*>  ")
+    out.append(summary())
+    out.append(if (!isSummaryComplete()) "..." else "")
+    out.append("  */")
+    out.result()
+  }
+
+  private def locatePreviousEdit(
+      statement: EvaluatedWorksheetStatement,
+      source: Input
+  ): Option[Position] = {
+    val editPattern = """\A\s*/\*>[\S\s]*?\*/""".r
+    val offset = source.lineToOffset(statement.position.endLine) + statement.position.endColumn
+    val text = source.text.drop(offset)
+    editPattern
+      .findFirstMatchIn(text)
+      .map(m => {
+        val p = source.toOffsetPosition(offset + m.end)
+        new Position(p.endLine, p.endColumn)
+      })
+  }
+
+  private def getHoverMessage(
+      position: Position,
+      hovers: Seq[HoverMessage]
+  ): Option[String] = {
+    hovers.find(_.range.encloses(position)).map(_.message)
+  }
+
+  private def updateWithEdits(text: String, edits: List[TextEdit]): String = {
+    val editsMap = edits.map(e => e.getRange().getStart().getLine() -> e).toMap
+
+    text.lines.zipWithIndex
+      .map {
+        case (line, i) =>
+          editsMap.get(i) match {
+            case Some(edit) =>
+              val before =
+                line.substring(0, edit.getRange.getStart.getCharacter)
+              val after = line.substring(edit.getRange.getEnd.getCharacter)
+              before + edit.getNewText() + after
+            case None => line
+          }
+      }
+      .mkString("\n")
+  }
+
+}
+
+object WorkspaceEditWorksheetPublisher {
+
+  case class HoverMessage(range: Range, message: String)
```

Comment:
let's move these to the toplevel, we're already in the worksheets package.

Reply:
but `RenderResult` does not really make sense outside of the class, don't you think?